### ISSUE TITLE
Running 4 MPI-dependent tasks in parallel on two compute nodes, how?

### ISSUE THREAD

**Issue by dz24** [radical-cybertools/radical.pilot#3237]
Hi, I am interested in using radical.pilot and it seems to work well for my purposes when it comes to running multiple independent Molecular Dynamics (MD) simulations on one node, within one HPC job. I have a question and hope this place is fine to ask it: I cannot seem to make the following situation work:

I want to run 4 tasks, e.g. "mpirun -np 32 cp2k", on two nodes having 64 cores each within one HPC job. The most efficient way to do this would be to run all 4 tasks in parallel, with one node working on two tasks at the same time. However, with different rp configurations, like using one or two pilots, I seem to get the same result of all four tasks seemingly being ran on one single node (despite having two nodes available), resulting in the performance being much worse than if I only ran a job with two tasks on one node instead. I am unable to see in the documentation or ways to make the individual tasks be exclusively distributed to individual nodes.

PBS script:

```bash
#!/bin/sh
#PBS -l select=2:ncpus=64:mpiprocs=64:ompthreads=1
#PBS -l walltime=00:10:00

# module version
module -s purge
module -s load cp2k/2024.2

# run radical pilot script
source ~/venvs/p39/bin/activate
python3 rp_script.py
```

radical pilot script:

```python3
import radical.pilot as rp
import time
import os

PWD = os.path.abspath(os.path.dirname(__file__))

# Create a session
session = rp.Session()

# Create a pilot manager and a pilot
pmgr    = rp.PilotManager(session=session)
pd_init = {'resource': 'local.localhost',
           'runtime' : 30,
           'cores'   : 128}
pdesc   = rp.PilotDescription(pd_init)
pilot   = pmgr.submit_pilots(pdesc)

# Crate a task manager and describe your tasks
tmgr = rp.TaskManager(session=session)
tmgr.add_pilots(pilot)
tds = list()
for i in range(4):
    td = rp.TaskDescription()
    td.executable     = '/apl/cp2k/2024.2/exe/rccs/cp2k.psmp'
    td.arguments      = [f'H2O-64.inp']
    td.ranks          = 32
    td.cores_per_rank = 1
    td.threading_type   = rp.OpenMP
    td.input_staging  = [f'{PWD}/H2O-64.inp']
    tds.append(td)
    print('submit', i)

# Submit your tasks for execution
tasks = tmgr.submit_tasks(tds)
start = time.perf_counter()

print(f'time taken: {time.perf_counter() - start:.02f}')
for task in tasks:
    print(f"{task.stdout.strip()} {task.stderr.strip()}")

# Close your session
session.close(cleanup=True)
```

localhost json file:
```json
{
    "localhost": {
        "description"                 : "Your local machine.",
        "default_queue"               : "H",
        "notes"                       : "To use the ssh schema, make sure that ssh access to localhost is enabled.",
        "default_schema"              : "local",
        "schemas"                     : {
            "ssh"                     : {
                "job_manager_endpoint": "ssh://localhost",
                "filesystem_endpoint" : "sftp://localhost"
            },
            "local"                   : {
                "job_manager_endpoint": "fork://localhost",
                "filesystem_endpoint" : "file://localhost"
            }
        },
        "default_remote_workdir"      : "$HOME",
        "resource_manager"            : "FORK",
        "agent_config"                : "default",
        "agent_scheduler"             : "CONTINUOUS",
        "agent_spawner"               : "POPEN",
        "launch_methods"              : {
                                         "order" : ["MPIRUN"],
                                         "MPIRUN" : {
                                             "pre_exec_cached": [
                                                 "module load 2024",
                                                 "module -s load cp2k/2024.2",
                                                 "module load cuda/12.2u2"
                                             ]
                                         }
                                        },
        "pre_bootstrap_0"             : ["module -s load cp2k/2024.2",
                                         "module load 2024",
                                         "module load cuda/12.2u2"],
        "rp_version"                  : "installed",
        "virtenv_mode"                : "local",
        "python_dist"                 : "default",
        "cores_per_node"              : 64,
        "gpus_per_node"               : 1,
        "lfs_path_per_node"           : "/tmp",
        "lfs_size_per_node"           : 1024,
        "mem_per_node"                : 504000,
        "fake_resources"              : true,
        "raptor"                      : {
                                         "hb_delay"     : 100,
                                         "hb_timeout"   : 400,
                                         "hb_frequency" : 100}
    }
}
```
 
This problem may be "trivial" with SLURM+srun , but now I am running on a PBS cluster which only offer mpirun. Is what I ask for possible with radical.pilot?

edit: Evidence for I believe all tasks end up being ran on one node, is, through grepping various rp files, all HOSTNAME refers to one node specifically, e.g. 

```
pilot.0001/env/lm_mpirun.sh:export HOSTNAME='node1'
```
and nothing comes up when searching for 'node2'. I imagine the problem could be fixed by modifying the -host input, which now is

```
pilot.0001/task.000003/task.000003.launch.sh:  mpirun  -np 32   -host localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost,localhost
```

**Comment by andre-merzky** [comment]
Hi @dz24, 

Yes, this is the right place to ask - thanks for opening the ticket!

You are overwriting the configuration for `local.localhost`.  While this is not the source of your problem, I still would like to suggest to instead to create a new json file (`resource_dz24.json` or so), or to at least add a new entry (`mycluster`) to `resource_local.json`.  

To the problem: RP is able to place your MPI tasks on the different nodes as you suggest.  However, to do so it needs to obtain information about what nodes are available to your specific job allocation.  That information is usually provided by the cluster's batch system which in your case is PBS.  However, the resource config you pasted above shows the `resource_manager` entry as `fork` - and that resource manager will only see the node the RP agent is running on, not the other nodes in your allocation.

So please change the `resource_manager` entry to `PBSPRO`, and let us know how that goes.

Best, Andre.

**Comment by dz24** [comment]
Hi, thanks for the quick reply!

I actually did try to set the `resource_manager` entry to `PBSPRO` at first, but for some reason my PBS cluster does not offer the qstat utility, which crashed the python script if i remember correctly. 

The cluster have a similar, but different utility called "jobinfo", which I could perhaps have fixed the crash in the code file `radical/pilot/agent/resource_manager/pbspro.py`, but I am not sure if it would have been an easy fix. Rather I think I made things work (for now) using mpirun with hostfiles and $PBS_NODEFILE.

I am not sure how much you could help in fixing this specific issue so I am fine with marking this ticket as solved, thanks again.

**Comment by andre-merzky** [comment]
I am glad that you got your code running :-)  Independent though, out of interest from our end:

> I am not sure how much you could help in fixing this specific issue...

we really don't mind covering additional batch system configurations, so support for `jobinfo` can well be added.  Could you meybe point us to some documentation pages for your cluster, mainly to check what pbs version is being used etc?  Thanks!

**Comment by dz24** [comment]
I am currently on the cluster: https://ccportal.ims.ac.jp/en/quickstartguide
https://ccportal.ims.ac.jp/en/manual/jobinfo

I am not sure if all information is in English or is documented, I just got access to it recently.

**Comment by andre-merzky** [comment]
Thanks for the info, @dz24.  Are you interested in getting this fixed for radical.pilot, or are you happy with your current approach?  If there is interest: what is the expected scale you expect to ultimately run your simulations on (how many nodes), and what clusters will you target eventually (or is this limited to the cluster you are on right now)?

**Comment by dz24** [comment]
Hi, 

I am fine with my current approach for now, though I have so far only run small tests and not any long-term production runs. On a previous cluster with SLURM, we were often fine with running a number of `srun` calls in parallel (until we frequently hit the job limit of 40k `srun` calls). From what I know `mpirun` does not have this limitation.

> what is the expected scale you expect to ultimately run your simulations on (how many nodes), and what clusters will you target eventually (or is this limited to the cluster you are on right now)?

It varies depending on the specific application that we are utilizing a specific [replica exchange method](https://github.com/infretis/infretis) to study. But as a ball-park estimate, between 1-10 nodes, where if possible we would fill 1-10 MD simulations per node depending on MD system and available hardware.

**Comment by andre-merzky** [comment]
Thank you for the information!  We could try to create a resource configuration for your cluster, but would probably need some time investment from your side also to test and provide feedback since we don't have accounts on your machine.  Would that approach work for you?


> limit of 40k srun calls

40k would be a great limit!  But I assume that should have been 40 srun calls?

**Comment by dz24** [comment]
> Would that approach work for you?

I may end up delaying running on that cluster, so it may not be necessarily just yet. But my group is applying for access to some other HPCs so I may reach out then, thanks for the offer.

> 40k would be a great limit! But I assume that should have been 40 srun calls?

It is (unnecessarily) a lot. It is actually a problem that I wanted to ask if you know any solutions for / if radical.pilot could help us solve. Basically we have e.g. 100k individual short MD simulations (tasks) that we want to run within one HPC job (or as many tasks as the requested wall time allows). Depending on how much hardware we can request from one HPC job, we decide that optimally X tasks should be performed/completed concurrently, where once a task is finished, the next task is started so that always X tasks are being ran during the HPC job.

Consequently you may understand how we ended up using more than 40K srun calls if e.g. 40 tasks were concurrently being ran on 40 individual node within one multi-node job, and every time a task completed we would have to re-allocate the freed up hardware to a new task via e.g. `srun -n 1 -N 1 -G 1 --exact md_program`. At that point I did not realize I could just have used `mpirun` with `hostfiles` instead, which does not have the SLURM overhead of a maximum limit? 

Is there therefore a way to allocate and split, the total amount of hardware requested within one HPC job to X "workers" and let them start/stop MD programs with their allocated fraction of hardware without invoking constantly mpirun/srun? Or is that what you imply is possible with

>  But I assume that should have been 40 srun calls?

**Comment by andre-merzky** [comment]
> Basically we have e.g. 100k individual short MD simulations (tasks) that we want to run within one HPC job...

Now, this is where your problem becomes interesting to us :-)  That is indeed the reason why RP was created in the first place: running large numbers of tasks within a single HPC job allocation.

So the answer is yes, we do support that use case, and in your specific case would indeed configure `mpirun` as launch method to execute the tasks to avoid the `srun` limit.

> Is there therefore a way to allocate and split, the total amount of hardware requested within one HPC job to X "workers" and let them start/stop MD programs with their allocated fraction of hardware without invoking constantly mpirun/srun? Or is that what you imply is possible with

It is certainly possible to implement node-level workers which run one MD after the other -  but RP also does that for you w/o the need to reimplement things.

Now, as said earlier, to get RP running on your machine, we do need to code some pieces in RP to work with your specific batch scheduler version.  Would you mind sending the output of the `jobinfo` command for a typical job allocation you are using?

**Comment by dz24** [comment]
Sorry for the late reply, at the moment I am only running simulations on my local machine, and do not require any queuing system. I'll let you know when/if large scale simulations are to be launched on hpcs.

**Comment by andre-merzky** [comment]
Great, that's what we do then!  I'll close this ticket meanwhile - please reopen at appropriate time, that will make it pop up on our TODO list :-)
